### PR TITLE
Added Polygon IHitShape

### PR DESCRIPTION
**Pull request by abcdefg30**
#13484 revived.

### CODE REVIEW

**Comment by Unit158** [comment]
Looks sane, and it works. I'm not fond of some of the variable naming, but I couldn't think of anything better so :+1:

**Review by Unit158 — APPROVED** [review]

File: OpenRA.Mods.Common/HitShapes/Polygon.cs [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+			OuterRadius = new WDist(Points.Max(x => x.Length));
+			combatOverlayVertsTop = Points.Select(p => new WVec(p.X, p.Y, VerticalTopOffset)).ToArray();
+			combatOverlayVertsBottom = Points.Select(p => new WVec(p.X, p.Y, VerticalBottomOffset)).ToArray();
+			squares = new int[Points.Length];
+			squares[0] = (Points[0] - Points[Points.Length - 1]).LengthSquared;
+			for (var i = 1; i < Points.Length; i++)
+				squares[i] = (Points[i] - Points[i - 1]).LengthSquared;
+		}
+
+		static int DistanceSquaredFromLineSegment(int2 c, int2 a, int2 b, int ab2)
```
Comment (reaperrr) [inline OpenRA.Mods.Common/HitShapes/Polygon.cs]: There must definitely be something wrong either in this...

File: OpenRA.Mods.Common/HitShapes/Polygon.cs [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+			// c is closest to its unknown orthogonal projection (p) onto the line spanned by b with a as the origin
+			var ab = b - a;
+			var ap2 = ac.X * ab.X + ac.Y * ab.Y;
+			var ap = new int2(ab.X * ap2 / ab2, ab.Y * ap2 / ab2);
+
+			// Length of vector pc squared.
+			return (ac - ap).LengthSquared;
+		}
+
+		public WDist DistanceFromEdge(WVec v)
```
Comment (reaperrr) [inline OpenRA.Mods.Common/HitShapes/Polygon.cs]: ...or this.

Shots that land rather close to an edge, but whose AoE does not include any of the polygons, deal zero damage.

Comment (Unit158) [reply OpenRA.Mods.Common/HitShapes/Polygon.cs]: Are you sure that the problem is in this code? I'm convinced it has to do with target scanning.

**Comment by Unit158** [comment]
I took a look at this a while back. I think there's at least an absolute value missing somewhere.

File: OpenRA.Mods.Common/HitShapes/Polygon.cs [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+			if (ac2 + ab2 <= bc2)
+				return ac2;
+
+			// c is closest to point b
+			if (bc2 + ab2 <= ac2)
+				return bc2;
+
+			// c is closest to its unknown orthogonal projection (p) onto the line spanned by b with a as the origin
+			var ab = b - a;
+			var ap2 = ac.X * ab.X + ac.Y * ab.Y;
```
Comment (Unit158) [inline OpenRA.Mods.Common/HitShapes/Polygon.cs]: This looks suspicious.

File: OpenRA.Mods.Common/HitShapes/Polygon.cs [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+			var ap = new int2(ab.X * ap2 / ab2, ab.Y * ap2 / ab2);
```
Comment (Unit158) [inline OpenRA.Mods.Common/HitShapes/Polygon.cs]: ab.X * ap^2 and ab.Y * ap^2 will overflow for larger hitboxes.

Comment (abcdefg30) [reply OpenRA.Mods.Common/HitShapes/Polygon.cs]: Fixed.

**Comment by reaperrr** [comment]
👍 after testcase has been removed.

**Comment by abcdefg30** [comment]
Removed it.

**Review by reaperrr — APPROVED** [review]

**Comment by reaperrr** [comment]
[changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/affc3145bd01fb5f2f095fdf3877696441063237...f9af8ebcb198b24d8883f88614954544e584197e)

**Pull request merged**